feature/deps.go

```diff
@@ -0,0 +1,109 @@
+package feature
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+type (
+	// featureSet contains a set of feature bits.
+	featureSet map[lnwire.FeatureBit]struct{}
+
+	// checkedFeatures maps a feature bit to a boolean indicating if this
+	// features depdencies have already been verified. This allows us to
+	// short circuit verification if multiple features have common
+	// depdencies, or map traversal starts verifying from the bottom up.
+	checkedFeatures map[lnwire.FeatureBit]bool
+
+	// depDesc maps a features to its set of dependent features, which must
+	// also be present for the vector to be valid. This can be used to
+	// recursively check the dependency chain for features in a feature
+	// vector.
+	depDesc map[lnwire.FeatureBit]featureSet
+)
+
+// ErrMissingFeatureDep is an error signaling that a transitive dependency in a
+// feature vector is not set properly.
+type ErrMissingFeatureDep struct {
+	dep lnwire.FeatureBit
+}
+
+// Error returns a human-readable description of the missing dep error.
+func (e ErrMissingFeatureDep) Error() string {
+	return fmt.Sprintf("missing feature: %v", e.dep)
+}
+
+// deps is the default set of dependencies for assigned feature bits. If a
+// feature is not present in the depDesc it is assumed to have no depdencies.
+var deps = depDesc{
+	lnwire.PaymentAddrOptional: {
+		lnwire.TLVOnionPayloadOptional: {},
+	},
+	lnwire.MPPOptional: {
+		lnwire.PaymentAddrOptional: {},
+	},
+}
+
+// ValidateDeps asserts that a feature vector sets all features and their
+// transitive dependencies properly. It assumes that the dependencies between
+// optional and required features are identical, e.g. if a feature is required
+// but its depdency is optional, that is sufficient.
+func ValidateDeps(fv *lnwire.FeatureVector) error {
+	features := fv.Features()
+	checked := initChecked(features)
+
+	return validateDeps(features, checked)
+}
+
+// validateDeps is a subroutine that recursively checks that the passed features
+// have all of their associated dependencies in the checked map.
+func validateDeps(features featureSet, checked checkedFeatures) error {
+	for bit := range features {
+		// Convert any required bits to optional.
+		if bit.IsRequired() {
+			bit ^= 0x01
+		}
+
+		// If the checked features doesn't contain the dependency, this
+		// vector is invalid. Alternatively, if we know that this
+		// dependencies are valid, we can short circuit and continue
+		// verifying other bits.
+		visited, ok := checked[bit]
+		if !ok {
+			return ErrMissingFeatureDep{bit}
```

Comment:
Maybe the name `checked` isn't sufficiently covered what it does. Besides checked, it also is a quick look up into the features that we have. I didn't immediately get that because of the name.